test/e2e/upgrade/upgrade.go

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) 2023 ScyllaDB.
+
+package upgrade
+
+import (
+	"context"
+	"regexp"
+	"strings"
+
+	"github.com/onsi/ginkgo/v2"
+	o "github.com/onsi/gomega"
+	"github.com/scylladb/scylla-operator/test/e2e/framework"
+)
+
+// Test is an interface for upgrade tests.
+type Test interface {
+	// Name should return a test name.
+	Name() string
+
+	// Setup should create and verify whatever objects need to
+	// exist before the upgrade disruption starts.
+	Setup(ctx context.Context, f *framework.Framework)
+
+	// Test will run during the upgrade. When the upgrade is
+	// complete, done will be closed and final validation can
+	// begin.
+	Test(ctx context.Context, f *framework.Framework, done <-chan struct{})
+
+	// Teardown should clean up any objects that are created that
+	// aren't already cleaned up by the framework. This will
+	// always be called, even if Setup failed.
+	Teardown(ctx context.Context, f *framework.Framework)
+}
+
+type UpgradeFunc func(ctx context.Context) error
+
+func RunUpgradeSuite(ctx context.Context, tests []Test, testFrameworks map[string]*framework.Framework, upgradeFunc UpgradeFunc) {
+	stopCh := make(chan struct{})
+	sems := make([]*semaphore, 0, len(tests))
+
+	for i := range tests {
+		tc := tests[i]
+
+		sem := &semaphore{
+			stop:  stopCh,
+			ready: make(chan struct{}, 1),
+			done:  make(chan struct{}, 1),
+		}
+		sems = append(sems, sem)
+
+		go func() {
+			defer ginkgo.GinkgoRecover()
+			defer sem.Done()
+
+			f := testFrameworks[tc.Name()]
+
+			ginkgo.DeferCleanup(tc.Teardown, f)
+
+			tc.Setup(ctx, f)
```

Comment:
Is my understanding correct that all these tests are going to run as a single Ginkgo `It` node? Couldn't they be ran as separate nodes in parallel? It seems more difficult to do, as it would require sharing memory between processes, but doesn't seem impossible either. This will get painful to debug once the set of tests grows.

Reply:
Yes in upgrade suite, single `It` validates how single disruption affects number of scenarios. They can't be run in parallel because we don't control synchronziation points in ginkgo. Why do you think debugging it's going to be painful? All tests have separate namespaces, you debug it after entire suite tears down. Feels the same to me.

Reply:
Fair, debugging won't be any messier, I was thinking about the convoluted logs in CI. If you started an upgrade in a `BeforeSuite` node, and implemented a barrier across processes, wouldn't that work? I guess the issue is you wouldn't have any control over when the tests are actually started with respect to the upgrade?

Reply:
> If you started an upgrade in a BeforeSuite node, and implemented a barrier across processes, wouldn't that work? 

Not really, `BeforeSuite` must end to start executing `It`. Tests need to set up resources before upgrade is started. We need to wait until all parallel setups are done, and only then introduce the disruption.

>  I guess the issue is you wouldn't have any control over when the tests are actually started with respect to the upgrade?

That's another issue.

Adding bidirectional multi process communication to implement barriers doesn't seem trivial. This approach is simple, at the cost of aggregated logs from single `It`. This can be fixed if framework would prefix log messages with the namespace it originates from.